File: bintree_java/src/binarytreedisp/DisplayParser.java
package binarytreedisp;


import binarytree.BinNode;



class DisplayParser {

//////////////////////////////////////////////////////////////
//                        FIELDS
//////////////////////////////////////////////////////////////



    public ValueUtil valueUtil = null;
    public char lineChar;
    public int lineBrsp;



//////////////////////////////////////////////////////////////
//                        METHODS
//////////////////////////////////////////////////////////////



    public DisplayParser(ValueUtil valueUtil) {
        this(valueUtil, '-', 1);
    }



    public DisplayParser(ValueUtil valueUtil, char character, int branchSpacing) {
        if (null == valueUtil)
            throw new IllegalArgumentException("valueUtil cannot be null");

        this.valueUtil = valueUtil;
        configLine(character, branchSpacing);
    }



    public void configLine(char character, int branchSpacing) {
        if (branchSpacing < 1)
            throw new IllegalArgumentException("branchSpacing must be a positive integer");

        this.lineChar = character;
        this.lineBrsp = branchSpacing;
    }



    public < TNode extends BinNode<?,TNode> >
    ParsingNode
    buildTree(TNode inputRoot)
    {
        if (null == inputRoot)
            return null;

        var node = new ParsingNode(this.valueUtil.getStr(inputRoot.key));
        int lenKey = node.key.length();

        node.left = buildTree(inputRoot.left);
        node.right = buildTree(inputRoot.right);

        boolean nodeLeftNull = (null == node.left);
        boolean nodeRightNull = (null == node.right);

        int widthLeftBranch = nodeLeftNull ? 0 : node.left.width;
        int widthRightBranch = nodeRightNull ? 0 : node.right.width;

        int sizeLeftLine = nodeLeftNull ? 0 : this.lineBrsp;
        int sizeRightLine = nodeRightNull ? 0 : this.lineBrsp;

        int fullWidth = widthLeftBranch + widthRightBranch + sizeLeftLine + sizeRightLine;

        int sizeRightOverflow = lenKey - (widthRightBranch + sizeRightLine);
        fullWidth += Math.max(1, sizeRightOverflow);

        int marginKey = widthLeftBranch + sizeLeftLine;
        int marginLeftChild = nodeLeftNull ? 0 : node.left.marginKey;
        int marginRightChild = nodeRightNull ? 0 : marginKey + node.right.marginKey + sizeRightLine + 1;

        node.width = fullWidth;
        node.widthLeftBranch = widthLeftBranch;
        node.widthRightBranch = widthRightBranch;
        node.sizeLeftLine = sizeLeftLine;
        node.sizeRightLine = sizeRightLine;

        node.marginKey = marginKey;
        node.marginLeftChild = marginLeftChild;
        node.marginRightChild = marginRightChild;

        return node;
    }



    public void destroyTree(ParsingNode node) {
        if (null == node)
            return;

        destroyTree(node.left);
        node.left = null;

        destroyTree(node.right);
        node.right = null;
    }



    public < TNode extends BinNode<?,TNode> >
    int
    getHeight(TNode node) {
        if (null == node)
            return 0;

        int heightLe = getHeight(node.left);
        int heightRi = getHeight(node.right);

        return 1 + Math.max(heightLe, heightRi);
    }
}

File: bintree_csharp/binarytreedisp/DisplayParser.cs
using System;
using my.binarytree;


namespace my.binarytreedisp
{
    class DisplayParser
    {

        //////////////////////////////////////////////////////////////
        //                        FIELDS
        //////////////////////////////////////////////////////////////



        public ValueUtil ValueUtil = null;
        public char LineChar;
        public int LineBrsp;



        //////////////////////////////////////////////////////////////
        //                        METHODS
        //////////////////////////////////////////////////////////////



        public DisplayParser(ValueUtil valueUtil, char character = '-', int branchSpacing = 1)
        {
            if (valueUtil is null)
                throw new ArgumentException("valueUtil cannot be null");

            this.ValueUtil = valueUtil;
            ConfigLine(character, branchSpacing);
        }



        public void ConfigLine(char character, int branchSpacing)
        {
            if (branchSpacing < 1)
                throw new ArgumentException("branchSpacing must be a positive integer");

            this.LineChar = character;
            this.LineBrsp = branchSpacing;
        }



        public ParsingNode BuildTree<TKey, TNode>(TNode inputRoot)
            where TKey : IComparable where TNode : BinNode<TKey, TNode>, new()
        {
            if (inputRoot is null)
                return null;

            var node = new ParsingNode(this.ValueUtil.GetStr(inputRoot.Key));
            int lenKey = node.Key.Length;

            node.Left = BuildTree<TKey, TNode>(inputRoot.Left);
            node.Right = BuildTree<TKey, TNode>(inputRoot.Right);

            bool nodeLeftNull = (node.Left is null);
            bool nodeRightNull = (node.Right is null);

            int widthLeftBranch = nodeLeftNull ? 0 : node.Left.Width;
            int widthRightBranch = nodeRightNull ? 0 : node.Right.Width;

            int sizeLeftLine = nodeLeftNull ? 0 : this.LineBrsp;
            int sizeRightLine = nodeRightNull ? 0 : this.LineBrsp;

            int fullWidth = widthLeftBranch + widthRightBranch + sizeLeftLine + sizeRightLine;

            int sizeRightOverflow = lenKey - (widthRightBranch + sizeRightLine);
            fullWidth += Math.Max(1, sizeRightOverflow);

            int marginKey = widthLeftBranch + sizeLeftLine;
            int marginLeftChild = nodeLeftNull ? 0 : node.Left.MarginKey;
            int marginRightChild = nodeRightNull ? 0 : marginKey + node.Right.MarginKey + sizeRightLine + 1;

            node.Width = fullWidth;
            node.WidthLeftBranch = widthLeftBranch;
            node.WidthRightBranch = widthRightBranch;
            node.SizeLeftLine = sizeLeftLine;
            node.SizeRightLine = sizeRightLine;

            node.MarginKey = marginKey;
            node.MarginLeftChild = marginLeftChild;
            node.MarginRightChild = marginRightChild;

            return node;
        }



        public void DestroyTree(ref ParsingNode node)
        {
            if (node is null)
                return;

            DestroyTree(ref node.Left);
            node.Left = null;

            DestroyTree(ref node.Right);
            node.Right = null;

            node.Key = null;
            node = null;
        }



        public int GetHeight<TKey, TNode>(TNode node)
            where TKey : IComparable where TNode : BinNode<TKey, TNode>, new()
        {
            if (node is null)
                return 0;

            int heightLe = GetHeight<TKey, TNode>(node.Left);
            int heightRi = GetHeight<TKey, TNode>(node.Right);

            return 1 + Math.Max(heightLe, heightRi);
        }
    }
}
